src/plugins/initiatives/discourse.js

```diff
@@ -0,0 +1,136 @@
+// @flow
+
+import {DomHandler, DomUtils, Parser} from "htmlparser2";
+
+import {type Topic, type Post} from "../discourse/fetch";
+import {type Initiative, type URL} from "./initiative";
+import {topicAddress} from "../discourse/address";
+
+type HeaderToURLsMap = Map<string, $ReadOnlyArray<URL>>;
+
+export async function initiativeFromDiscourseTracker(
+  serverUrl: string,
+  topic: Topic,
+  firstPost: Post
+): Promise<Initiative> {
+  if (firstPost.topicId !== topic.id) {
+    throw new Error("Post is from a different topic");
+  }
+
+  if (firstPost.indexWithinTopic !== 1) {
+    throw new Error("Post is not the first post in the topic");
+  }
+
+  const {title} = topic;
+  const {timestampMs} = firstPost;
+  const tracker = topicAddress(serverUrl, topic.id);
+
+  const map = await groupURLsByHeader(firstPost.cooked);
+  const status = findStatus(map);
+  const champions = singleMatch(map, new RegExp(/^Champions?/i));
+  const contributions = singleMatch(map, new RegExp(/^Contributions?/i));
+  const dependencies = singleMatch(map, new RegExp(/^Dependenc(y|ies)/i));
+  const references = singleMatch(map, new RegExp(/^References?/i));
+
+  const missing = [];
+  if (status === null) missing.push("status");
+  if (!champions) missing.push("champions");
+  if (!dependencies) missing.push("dependencies");
+  if (!references) missing.push("references");
+  if (!contributions) missing.push("contributions");
+
+  if (missing.length > 0) {
+    const missingStr = JSON.stringify(missing);
+    throw new Error(
+      `Missing or malformed headers ${missingStr} for initiative topic "${title}" (${topic.id})`
+    );
+  }
+
+  const completedRE = new RegExp(/^completed?$/i);
+  const completed = status ? completedRE.test(status.trim()) : false;
+
+  return {
+    title,
+    tracker,
+    timestampMs,
+    completed,
+    champions: champions || [],
```

Comment:
I disagree on this. There's a clear benefit here which flow does not handle in a pretty way.
I think having a list of things we did wrong is a lot more valuable than a single thing we did wrong and having to solve those problems one at a time because it failed to inform us all at once.

This is purely a coercion to convince flow it's never null at this point.
If there's a better way to cast the type to be not nullable here, I'd love to hear.
As we've checked for this with `if (missing.length > 0)`.

The same applies to `const completed = status ? completedRE.test(status.trim()) : false;` as status is not null here, this is a completely redundant if-statement when we know status is definitely a string here.

Reply:
Unless you meant in the same way that I did for the errors of this function.
Instead of `Thing | null` return types. Have it be `Thing` or throw. Catch those and throw the collection of errors.

Reply:
I found the only other way to do this seems to be following flows' stricter type refinement approach. https://flow.org/en/docs/lang/refinements/#toc-refinement-invalidations

Replacing
```
if (missing.length > 0)
```

with 
```
if (
    completed == null ||
    champions == null ||
    dependencies == null ||
    references == null ||
    contributions == null
  )
```
even though it's redundant.

Reply:
I've implemented the later option, with the large if statement to do flow type refinement.